File: integration-tests/node-openai/tests/basic.test.js
import { test, describe } from 'node:test';
import assert from 'node:assert';
import OpenAI from 'openai';

// Configuration
const TEENYTINY_URL = process.env.TEENYTINY_URL || 'http://localhost:8080';
const TEENYTINY_API_KEY = process.env.TEENYTINY_API_KEY || 'testkey';

// Create OpenAI client configured for TeenyTiny AI
const openai = new OpenAI({
  apiKey: TEENYTINY_API_KEY,
  baseURL: `${TEENYTINY_URL}/v1`,
});

describe('Basic OpenAI SDK Integration', () => {
  test('basic completion with echo model', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [
        { role: 'user', content: 'Hello World' }
      ],
    });

    assert.strictEqual(completion.choices[0].message.content, 'Hello World');
    assert.strictEqual(completion.model, 'echo');
    assert.strictEqual(completion.object, 'chat.completion');
    assert.ok(completion.usage.total_tokens > 0);
  });

  test('multi-message conversation (should return last user message)', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [
        { role: 'system', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'First message' },
        { role: 'assistant', content: 'First response' },
        { role: 'user', content: 'Last message' }
      ],
    });

    assert.strictEqual(completion.choices[0].message.content, 'Last message');
  });

  test('system prompt with user message', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [
        { role: 'system', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'Test message' }
      ],
    });

    // Echo model should return the user message, ignoring system prompt
    assert.strictEqual(completion.choices[0].message.content, 'Test message');
  });

  test('no user messages (system only) returns default response', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [
        { role: 'system', content: 'You are a helpful assistant.' }
      ],
    });

    // Should get the default echo model greeting
    const content = completion.choices[0].message.content;
    assert.ok(content.includes('Echo model'), `Expected default response, got: ${content}`);
  });

  test('response structure matches OpenAI format', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [{ role: 'user', content: 'Structure test' }],
    });

    // Check required fields
    assert.ok(completion.id);
    assert.strictEqual(completion.object, 'chat.completion');
    assert.ok(completion.created);
    assert.strictEqual(completion.model, 'echo');
    assert.ok(Array.isArray(completion.choices));
    assert.ok(completion.usage);

    // Check choice structure
    const choice = completion.choices[0];
    assert.strictEqual(choice.index, 0);
    assert.ok(choice.message);
    assert.strictEqual(choice.finish_reason, 'stop');

    // Check message structure
    assert.strictEqual(choice.message.role, 'assistant');
    assert.strictEqual(choice.message.content, 'Structure test');

    // Check usage structure
    assert.ok(typeof completion.usage.prompt_tokens === 'number');
    assert.ok(typeof completion.usage.completion_tokens === 'number');
    assert.ok(typeof completion.usage.total_tokens === 'number');
  });

  test('empty message handling', async () => {
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [{ role: 'user', content: '' }],
    });

    // Empty input should be handled gracefully
    assert.ok(typeof completion.choices[0].message.content === 'string');
  });

  test('special characters and unicode', async () => {
    const testMessage = 'Hello! 🌟 Special chars: @#$%^&*()_+-={}[]|\\:;"\'<>?,./ 中文';
    
    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [{ role: 'user', content: testMessage }],
    });

    assert.strictEqual(completion.choices[0].message.content, testMessage);
  });

  test('multiline content', async () => {
    const multilineMessage = `Line 1
Line 2
Line 3 with more content
Final line`;

    const completion = await openai.chat.completions.create({
      model: 'echo',
      messages: [{ role: 'user', content: multilineMessage }],
    });

    assert.strictEqual(completion.choices[0].message.content, multilineMessage);
  });
});

File: integration-tests/go-openai/basic_test.go
package main

import (
	"context"
	"os"
	"testing"

	"github.com/sashabaranov/go-openai"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func setupClient(t *testing.T) *openai.Client {
	baseURL := os.Getenv("TEENYTINY_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}

	apiKey := os.Getenv("TEENYTINY_API_KEY")
	if apiKey == "" {
		apiKey = "testkey"
	}

	config := openai.DefaultConfig(apiKey)
	config.BaseURL = baseURL + "/v1"

	return openai.NewClientWithConfig(config)
}

func TestBasicCompletion(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "Hello World",
				},
			},
		},
	)

	require.NoError(t, err)
	assert.Equal(t, "Hello World", resp.Choices[0].Message.Content)
	assert.Equal(t, "echo", resp.Model)
	assert.Equal(t, "chat.completion", resp.Object)
	assert.Greater(t, resp.Usage.TotalTokens, 0)
}

func TestMultiMessageConversation(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: "You are a helpful assistant.",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "First message",
				},
				{
					Role:    openai.ChatMessageRoleAssistant,
					Content: "First response",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "Last message",
				},
			},
		},
	)

	require.NoError(t, err)
	assert.Equal(t, "Last message", resp.Choices[0].Message.Content)
}

func TestSystemPromptWithUserMessage(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: "You are a helpful assistant.",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "Test message",
				},
			},
		},
	)

	require.NoError(t, err)
	// Echo model should return the user message, ignoring system prompt
	assert.Equal(t, "Test message", resp.Choices[0].Message.Content)
}

func TestSystemOnlyReturnsDefault(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: "You are a helpful assistant.",
				},
			},
		},
	)

	require.NoError(t, err)
	// Should get the default echo model greeting
	assert.Contains(t, resp.Choices[0].Message.Content, "Echo model")
}

func TestResponseStructure(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "Structure test",
				},
			},
		},
	)

	require.NoError(t, err)

	// Check required fields
	assert.NotEmpty(t, resp.ID)
	assert.Equal(t, "chat.completion", resp.Object)
	assert.Greater(t, resp.Created, int64(0))
	assert.Equal(t, "echo", resp.Model)
	assert.NotEmpty(t, resp.Choices)

	// Check choice structure
	choice := resp.Choices[0]
	assert.Equal(t, 0, choice.Index)
	assert.NotEmpty(t, choice.Message)
	assert.Equal(t, openai.FinishReasonStop, choice.FinishReason)

	// Check message structure
	assert.Equal(t, openai.ChatMessageRoleAssistant, choice.Message.Role)
	assert.Equal(t, "Structure test", choice.Message.Content)

	// Check usage structure
	assert.GreaterOrEqual(t, resp.Usage.PromptTokens, 0)
	assert.GreaterOrEqual(t, resp.Usage.CompletionTokens, 0)
	assert.Greater(t, resp.Usage.TotalTokens, 0)
}

func TestEmptyMessageHandling(t *testing.T) {
	client := setupClient(t)

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: "",
				},
			},
		},
	)

	require.NoError(t, err)
	// Empty input should be handled gracefully
	assert.NotNil(t, resp.Choices[0].Message.Content)
}

func TestSpecialCharactersAndUnicode(t *testing.T) {
	client := setupClient(t)
	testMessage := "Hello! 🌟 Special chars: @#$%^&*()_+-={}[]|\\:;\"'<>?,./ 中文"

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: testMessage,
				},
			},
		},
	)

	require.NoError(t, err)
	assert.Equal(t, testMessage, resp.Choices[0].Message.Content)
}

func TestMultilineContent(t *testing.T) {
	client := setupClient(t)
	multilineMessage := "Line 1\nLine 2\nLine 3 with more content\nFinal line"

	resp, err := client.CreateChatCompletion(
		context.Background(),
		openai.ChatCompletionRequest{
			Model: "echo",
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleUser,
					Content: multilineMessage,
				},
			},
		},
	)

	require.NoError(t, err)
	assert.Equal(t, multilineMessage, resp.Choices[0].Message.Content)
}